app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,80 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def openid_connect
+    return error_redirect unless valid_params?
+
+    session.delete(:gov_one_auth_state)
+
+    auth_service = GovOneAuthService.new(params['code'])
+    tokens_response = auth_service.tokens
+
+    return error_redirect unless valid_tokens?(tokens_response)
+
+    id_token = tokens_response['id_token']
+    session[:id_token] = id_token
+    gov_one_id = auth_service.decode_id_token(id_token)[0]['sub']
+
+    user_info_response = auth_service.user_info(tokens_response['access_token'])
+    return error_redirect unless valid_user_info?(user_info_response)
+
+    gov_user = find_or_create_user(user_info_response['email'], gov_one_id)
+    sign_in_and_redirect gov_user if gov_user
+  end
+
+private
+
+  # @return [Boolean]
+  def valid_params?
+    params['code'].present? && params['state'].present? && params['state'] == session[:gov_one_auth_state]
+  end
+
+  # @param tokens_response [Hash]
+  # @return [Boolean]
+  def valid_tokens?(tokens_response)
+    tokens_response.present? && tokens_response['access_token'].present? && tokens_response['id_token'].present?
+  end
+
+  # @param user_info_response [Hash]
+  # @return [Boolean]
+  def valid_user_info?(user_info_response)
+    user_info_response.present? && user_info_response['email'].present?
+  end
+
+  # @return [void]
+  def error_redirect
+    flash[:alert] = 'There was a problem signing in. Please try again.'
+    redirect_to root_path
+  end
+
+  # @return [User]
+  def find_or_create_user(email, id)
```

Comment:
I feel like this method might come out a little cleaner if it was given some more attention.

Reply:
Ah yes, that was messy! [773178d](https://github.com/DFE-Digital/early-years-foundation-recovery/pull/917/commits/773178de60d063b1e23b023cab9470aec05273c7)

Reply:
What would happen if we gave the responsibility of all this to a `User.find_or_create_from_gov_one` method?